File: src/gestorAplicación/modelos/Tienda.java
package gestorAplicacion.modelos;

import java.io.Serializable;
import java.util.*;

import gestorAplicacion.operaciones.Venta;

public class Tienda extends Establecimiento implements Serializable {
    private static final long serialVersionUID = 1L;

    private static HashMap<Integer, TipoProducto> descuentos;
    private ArrayList<Producto> productos = new ArrayList<>();
    private ArrayList<Venta> ventas = new ArrayList<>();
    private ArrayList<Cliente> clientes = new ArrayList<Cliente>();
    private ArrayList<Bodega> bodegas = new ArrayList<Bodega>();
    private int tamanoAlmacenamiento;
    private int minProducto;
    private Date fechaHoy;

    static {
        descuentos = new HashMap<>();
        descuentos.put(1, TipoProducto.Alimentos);
        descuentos.put(2, TipoProducto.Aseo);
        descuentos.put(3, TipoProducto.Hogar);
        descuentos.put(4, TipoProducto.Confiteria);
        descuentos.put(5, TipoProducto.Licores);
    }

    public Tienda() {
        super();
    }

    public Tienda(int tamanoAlmacenamiento, int minProducto, Date fechaHoy) {
        super(tamanoAlmacenamiento, minProducto);
        this.fechaHoy = fechaHoy;
    }

    // cacula la suma total de cantidad por producto y se la resta al
    // tamanoAlmacenamiento.
    public int verDisponibilidad() {
        int cantidadOcupada = 0;
        for (Producto producto : productos) {
            cantidadOcupada = producto.getCantidad();
        }
        return (tamanoAlmacenamiento - cantidadOcupada);
    }

    // como se genera la venta desde la clase Tienda??
    public Venta generarVenta(Date fecha, Cliente cliente, ArrayList<Producto> productos) {
        Venta nuevaVenta = new Venta(fecha, productos, cliente);
        ventas.add(nuevaVenta);
        cliente.agregarCompra(nuevaVenta);

        for (Producto productoVenta : productos) {
            Producto productoEnStock = buscarProductoPorID(productoVenta.getId());
            int cantidadProductoEnStock = productoEnStock.getCantidad();
            int cantidadProductoEnVenta = productoVenta.getCantidad();
            if ((cantidadProductoEnStock - cantidadProductoEnVenta) < 0) {
                pedirTraslado(productoEnStock, Math.abs(cantidadProductoEnStock-cantidadProductoEnVenta));
            }
            productoEnStock.setCantidad(cantidadProductoEnStock - cantidadProductoEnVenta);
        }

        return nuevaVenta;
    }


    // Seria bueno que retorne un Array con los productos con stock bajo??
    public void alertaStockBajo() {
        ArrayList<Producto> productosBajoStock = new ArrayList<Producto>();
        for (Producto producto : productos) {
            if (producto.getCantidad() < minProducto) {
                productosBajoStock.add(producto);
            }
        }
    }

    public void pedirTraslado(Producto producto, int cantidad) {
        /*
        HashMap<Bodega, Integer> bodegaIndicada = new HashMap<>();
        bodegaIndicada.put(null, 0);
        
        for (Bodega bodg : bodegas){
            Producto productoBodega = bodg.buscarProductoPorID(producto.getId());
            if (productoBodega != null && productoBodega.getCantidad() > (int) bodegaIndicada.values().toArray()[0]) bodegaIndicada.put(bodg,productoBodega.getCantidad());
        }
        */

        Bodega bodegaIndicada = null;
        int idProducto = producto.getId();
        
        for(Bodega bodg : bodegas){
            if (bodegaIndicada == null) bodegaIndicada = bodg;
            else if (bodg.buscarProductoPorID(idProducto).getCantidad() > bodegaIndicada.buscarProductoPorID(idProducto).getCantidad()) bodegaIndicada = bodg;
        }

        bodegaIndicada.trasladoTienda(idProducto, cantidad);
        producto.setCantidad(producto.getCantidad()+cantidad);
    }

    public void ofertaDia(ArrayList<Producto> productosVenta, Date fecha) {
        Iterator<Producto> itProd = productosVenta.iterator();
        while (itProd.hasNext()) {
            Producto tempProd = itProd.next();
            if (tempProd.getTipoProducto() == descuentos.get(fecha.getDate())) {
                double descuento = Math.random() * 0.05;
                tempProd.setDescuento(descuento);
            }
        }
    }

    public void agregarCliente(Cliente cliente) {
        clientes.add(cliente);
    }

    public void agregarBodega(Bodega bodega) {
        bodegas.add(bodega);
    }

    public Cliente buscarClientePorIdentificacion(String documento) {
        Iterator<Cliente> itClien = clientes.iterator();
        while (itClien.hasNext()) {
            Cliente tempClien = itClien.next();
            if (tempClien.getDocumento().equals(documento)) {
                return tempClien;
            }
        }
        return null;
    }

    public Cliente buscarClientePorNombre(String nombre) {
        Iterator<Cliente> itClien = clientes.iterator();
        while (itClien.hasNext()) {
            Cliente tempClien = itClien.next();
            if (tempClien.getNombre().equals(nombre)) {
                return tempClien;
            }
        }
        return null;
    }

    @Override
    public int getMinProducto() {
        // TODO Auto-generated method stub
        return super.getMinProducto();
    }

    @Override
    public void setMinProducto(int minProducto) {
        // TODO Auto-generated method stub
        super.setMinProducto(minProducto);
    }

    @Override
    public int getTamanoAlmacenamiento() {
        // TODO Auto-generated method stub
        return super.getTamanoAlmacenamiento();
    }

    public Date getFechaHoy() {
        return fechaHoy;
    }

    public void setFechaHoy(Date fechaHoy) {
        this.fechaHoy = fechaHoy;
    }

    public ArrayList<Venta> getVentas() {
        return ventas;
    }

    public void setVentas(Venta venta) {
        this.ventas.add(venta);
    }

    @Override
    public ArrayList<Producto> getProductos() {
        // TODO Auto-generated method stub
        return super.getProductos();
    }

    public void setProductos(ArrayList<Producto> productos) {
        this.productos = productos;
    }

    public ArrayList<Cliente> getClientes() {
        return clientes;
    }

    public void setClientes(ArrayList<Cliente> clientes) {
        this.clientes = clientes;
    }

    public ArrayList<Bodega> getBodegas() {
        return bodegas;
    }

    public void setBodegas(ArrayList<Bodega> bodegas) {
        this.bodegas = bodegas;
    }

}

File: python/gestorAplicacion/modelos/Tienda.py
from gestorAplicacion.modelos.Establecimiento import Establecimiento
from gestorAplicacion.operaciones.Venta import Venta 
from gestorAplicacion.modelos.TipoProducto import TipoProducto
from gestorAplicacion.modelos.Cliente import Cliente
from gestorAplicacion.modelos.Producto import Producto
from gestorAplicacion.modelos.Bodega import Bodega
import datetime as dt
from random import random as rd

class Tienda(Establecimiento):

    _descuentos = {TipoProducto.Alimentos: 0, TipoProducto.Aseo: 1,
                 TipoProducto.Hogar: 2, TipoProducto.Confiteria: 3, TipoProducto.Licores: 4}

    
    def __init__(self, tamanoAlmacenamiento=0, minProducto=0, fechahoy=None):
        super().__init__()
        self._fechahoy = fechahoy
        self._ventas = []
        self._clientes = []
        self._bodegas = []

    
    def getFechaHoy(self):
        return self._fechahoy
    
    def setFechaHoy(self, fecha):
        self._fechahoy = fecha

    def getVentas(self):
        return self._ventas
    
    def setVentas(self, ventas):
        self._ventas = ventas

    def getClientes(self):
        return self._clientes
    
    def setClientes(self, clientes):
        self._clientes = clientes

    def getBodegas(self):
        return self._bodegas
    
    def setBodegas(self, bodegas):
        self._bodegas = bodegas

    def agregarVenta(self, venta):
        self._ventas.append(venta)

    def agregarCliente(self, cliente):
        self._clientes.append(cliente)

    def generarVenta(self, fecha, cliente, productos):
        
        nuevaVenta = Venta(fecha, productos, cliente)

        self._ventas.append(nuevaVenta)
        cliente.agregarOperacion(nuevaVenta)

        for prodVenta in productos:
            productoEnStock = self.buscarProductoPorIDOriginal(prodVenta.getId())
            cantidadProductoEnStock = productoEnStock.getCantidad()
            cantidadProductoEnVenta = prodVenta.getCantidad()

            if (cantidadProductoEnStock-cantidadProductoEnVenta) < 0:
                self.pedirTraslado(fecha, productoEnStock, abs(cantidadProductoEnStock-cantidadProductoEnVenta))

            productoEnStock.setCantidad(productoEnStock.getCantidad()-cantidadProductoEnVenta)

        return nuevaVenta
    

    def productoStockBajo(self, producto):
        if producto.getCantidad() < self.getMinProducto():
            return True
        else:
            return False
        
    def cantidadEnBodegas(self, producto):
        cantidadTotal = 0
        idProdcuto = producto.getId()
        for bodega in self._bodegas:
            for producto4 in bodega.getProductos():
                if producto4.getId() == idProdcuto:
                    cantidadTotal += producto4.getCantidad()
        
        return cantidadTotal
        
    def pedirTraslado(self, fecha, producto, cantidad):
        
        prodVerificar = producto
        bodegaIndicada = None
        idProducto = producto.getId()
        
        for bodg in self._bodegas:
            if bodegaIndicada == None:
                bodegaIndicada = bodg
            elif bodg.buscarProductoPorIDOriginal(idProducto) != None and bodegaIndicada.buscarProductoPorIDOriginal(idProducto) != None:
                if bodg.buscarProductoPorIDOriginal(idProducto).getCantidad() > bodegaIndicada.buscarProductoPorIDOriginal(idProducto).getCantidad():
                    bodegaIndicada = bodg
            else:
                bodegaIndicada = bodg if bodg.buscarProductoPorIDOriginal(idProducto) != None else bodegaIndicada

        if self.buscarProductoPorIDOriginal(prodVerificar.getId()) != None:
            bodegaIndicada.trasladoTienda(fecha, idProducto, cantidad)
            producto.setCantidad(producto.getCantidad() + cantidad)
        else:
            bodegaIndicada.trasladoTienda(fecha, idProducto, cantidad)
            self.agregarProducto(producto)

    
    def ofertaDia(self, productosVenta, fecha):
        tipoDescuento = None
        for key, value in Tienda._descuentos.items():
            if value == fecha.weekday():
                tipoDescuento = key

        for prodVent in productosVenta:
            if prodVent.getTipoProducto() == tipoDescuento:
                descuento = round(rd() * 0.05 * 100.0) / 100.0
                prodVent.setDescuento(descuento)

    def agregarCliente(self, cliente):
        self._clientes.append(cliente)

    def buscarClientePorIdentificacion(self, documento):
        for cliente in self._clientes:
            if cliente.getDocumento() == documento:
                return cliente



    def buscarClientePorNombre(self, nombre):
        for cliente in self._clientes:
            if cliente.getNombre() == nombre:
                return cliente

